sycl/plugins/unified_runtime/ur/adapters/level_zero/usm.cpp

```diff
@@ -810,19 +935,14 @@ ur_result_t USMFreeHelper(ur_context_handle_t Context, void *Ptr,
 
   // If memory type is host release from host pool
   if (ZeMemoryAllocationProperties.type == ZE_MEMORY_TYPE_HOST) {
-    try {
-      Context->HostMemAllocContext->deallocate(Ptr);
-    } catch (const UsmAllocationException &Ex) {
-      return Ex.getError();
-    } catch (...) {
-      return UR_RESULT_ERROR_UNKNOWN;
-    }
+    auto umfRet = umfPoolFree(Context->HostMemPool.get(), Ptr);
```

Comment:
there doesn't seem to be a path where we determine that the pointer was allocated with a `ur_usm_pool_handle_t` passed by the user and then use the appropriate pool from that object instead of one that lives in `Context` for this call, do we need something like that?

Reply:
if we do it might be worth looking at adding an optional pool argument to the spec for `urUSMFree` to simplify that

Reply:
Actually, in UMF we have 2 functions for free:
`umfPoolFree(umf_memory_pool_handle_t hPool, void *ptr)`
`umfFree(void *ptr)`

Here, we could use the second one which automatically figures out the pool. However, this requires tracking all coarse-grain (driver) allocations in umf and might introduce some overhead. In the long-term, I don't think this overhead would be significant but for now, our implementation is not yet optimized and we're actually evaluating if there is any performance impact.

Adding an extra argument to `urUSMFree` might be a good idea to avoid this overhead. However, what if the user passes `NULL` as `ur_usm_pool_handle_t`? That could mean that he didn't use any explicit pool or that he doesn't keep track of it - we would still need to do the lookup (and that would be the most common scenario I think). Perhaps it would be better to just have 2 functions, just like in UMF?

Reply:
Btw, I think it might be better to fix the free operation for allocations made from an explicit pool in a separate PR. since it might require some refactoring - there is some code for looking up the pools that might no longer be necessary.

Also, if we'll use the umf tracking capabilities,  we can also leverage it in urUSMGetMemAllocInfo and possibly other places to avoid calling L0.

Reply:
Ok cool, I didn't notice that umf does some tracking of its own - I was expecting the adapter to have to do that.

I would expect the new pool parameter to be an optional fast path, like the user can skip the pointer lookup by giving us the correct pool. Not necessary right now I guess but it's an option if we're looking for ways to shave overhead in the future.